pkg/kubernetes/reconcile.go

```diff
@@ -14,138 +14,129 @@ import (
+	// our kubernetes specification
+	for _, manifest := range manifests {
+		if spec.Namespace != "" && !manifest.Metadata().HasNamespace() {
+			manifest.Metadata()["namespace"] = spec.Namespace
 		}
-
-		out = append(out, m)
 	}
 
-	// optionally filter the working set of objects
-	if len(targets) > 0 {
-		tmp := funk.Filter(out, func(i interface{}) bool {
-			p := objectspec(i.(manifest.Manifest))
-			for _, t := range targets {
-				if t.MatchString(strings.ToLower(p)) {
+	// If we have any kind-name matchers, we should filter all the manifests by matching
+	// against their <kind>/<name> identifier.
+	if len(kindNameMatchers) > 0 {
+		manifests = funk.Filter(manifests, func(elem interface{}) bool {
+			manifest := elem.(Manifest)
+			kindName := strings.ToLower(manifest.KindName())
+			for _, matcher := range kindNameMatchers {
+				if matcher.MatchString(kindName) {
 					return true
 				}
 			}
+
 			return false
-		}).([]manifest.Manifest)
-		out = manifest.List(tmp)
+		}).([]Manifest)
 	}
 
-	// Stable output order
-	sort.SliceStable(out, func(i int, j int) bool {
-		if out[i].Kind() != out[j].Kind() {
-			return out[i].Kind() < out[j].Kind()
-		}
-		return out[i].Metadata().Name() < out[j].Metadata().Name()
+	sort.SliceStable(manifests, func(i int, j int) bool {
+		return manifests[i].KindName() < manifests[j].KindName()
 	})
 
-	return out, nil
+	return manifests, nil
 }
 
-func extract(deep interface{}) (map[string]manifest.Manifest, error) {
-	extracted := make(map[string]manifest.Manifest)
-	if err := walkJSON(deep, extracted, nil); err != nil {
-		return nil, err
-	}
-	return extracted, nil
+type ErrorPrimitiveReached struct {
+	path      string
+	primitive interface{}
 }
 
-// walkJSON traverses deeply nested kubernetes manifest and extracts them into a flat []dict.
-func walkJSON(deep interface{}, extracted map[string]manifest.Manifest, path trace) error {
-	// array: walkJSON for each
-	if d, ok := deep.([]map[string]interface{}); ok {
-		for i, j := range d {
-			path := append(path, fmt.Sprintf("[%v]", i))
-			if err := walkJSON(j, extracted, path); err != nil {
-				return err
-			}
-		}
-		return nil
-	}
+func (e ErrorPrimitiveReached) Error() string {
+	return fmt.Sprintf("found an object at %s of type %T that was not a valid Kubernetes object", e.path, e.primitive)
+}
 
-	// assert for map[string]interface{} (also aliased objx.Map)
-	if m, ok := deep.(objx.Map); ok {
-		deep = map[string]interface{}(m)
+// walkJSON recurses into either a map or list, returning a list of all objects that look
+// like kubernetes resources. We support resources at an arbitrary level of nesting, and
+// return an error if any leaf nodes f
+//
+// Handling the different types is quite gross, so we split this method into a generic
+// walkJSON, and then walkObj/walkList to handle the two different types of collection we
+// support.
+func walkJSON(ptr interface{}, paths ...string) ([]Manifest, error) {
```

Comment:
I'd prefer to leave the argument called `deep`, because `ptr` doesn't tell at all what it is (**deep**ly nested manifests)

Furthermore, please return `map[string]manifest.Manifest` here as it was before, because that allows to test that the `jsonpath` building works correctly

Reply:
> I'd prefer to leave the argument called deep, because ptr doesn't tell at all what it is (deeply nested manifests)

That's interesting, as whatever we call this variable-deep or ptr or whatever- it's not a deeply nested object, right?

This function has to handle the case where it's either an object, a slice, or something else entirely. So calling it deep, where you're trying to say that deep means "deeply nested object", would be wrong in all but one of the several cases we rely on, right?

Reply:
You got the point there :+1: